### PR TITLE
Softdevice bootloader dfu

### PR DESCRIPTION
**Pull request by mjdietzx**
Now either the SoftDevice, the application or the bootloader can be updated via DFU OTA.

Modified hex2bin to convert hex files to binaries. It now allows gaps in the hex file and will pad the binary with 0xFF. It also allows us to update the Softdevice (by trimming off the master boot record data from the binary as we never update this) and the bootloader (by trimming off any data that writes to the UICR). Both of these are as described in Nordic's DFU spec.

Tested and working with application, softdevice and bootloader updates. The user will have to specify minAddress if updating softdevice which is the end of the master boot record, and both maxAddress if updating bootloader (which is the last address in flash).

Note: Hex converter is a bit strict and throws errors for invalid hex files. based off of https://en.wikipedia.org/wiki/Intel_HEX. Also updating the softdevice and bootloader at the same time will not work right now. Some hex files have been added for testing DFU of SD and BL.

### CODE REVIEW

**Comment by thegecko** [comment]
@mjdietzx Awesome work!

Could you update the README to describe how to use the new hex file?

**Comment by mjdietzx** [comment]
thankyou! yes good idea.

**Comment by thegecko** [comment]
Thanks, I was hoping the README additions could reference the new `nrf52_bootloader_only.hex` firmware file to allow download and describe what it is.

Unless it isn't needed?

**Comment by mjdietzx** [comment]
OK, I will add this description to the README. I think it is good to have - the bootloader_only.hex and softdevice_only.hex allow us to do a DFU of the bootloader or softdevice respectively. These will be good for testing

**Comment by mjdietzx** [comment]
Is this what you were looking for?

Do you think I have too much info. in the README? Not sure what you intend here but we could probably get away with only:

'The .hex file below can be transferred via DFU to test bootloader updates.
nrf52_bootloader_only.hex. Nordic's standard bootloader (nRF52 only).'

for both the SoftDevice and Bootloader. I could take out all that info on Master Boot Record and UICR if you don't want it in the README

**Comment by thegecko** [comment]
It is a bit wordy TBH :) If this documentation is hosted elsewhere (nordic SDK for example) I'd be happier linking to it so it doesn't get out of date.

**Comment by mjdietzx** [comment]
Sorry about all these commits... Gonna make one more.. I think I need to find a better place to document this stuff. Just going to explain the new .hex files as you suggest

**Comment by thegecko** [comment]
Thanks, let me know when you are happy

**Comment by mjdietzx** [comment]
I'm happy with it now as long as you don't see anything else that could be improved.
